File: app/schemas/liability.py
from datetime import date
from decimal import Decimal

from pydantic import BaseModel

from app.schemas.account import ORMModel


class LiabilityResponse(ORMModel):
    id: int
    type: str
    name: str
    institution: str | None
    original_amount: Decimal
    current_balance: Decimal
    interest_rate: Decimal | None
    due_day: int | None
    notes: str | None
    is_active: bool


class LiabilityCreate(BaseModel):
    type: str
    name: str
    institution: str | None = None
    original_amount: Decimal = Decimal("0")
    current_balance: Decimal
    interest_rate: Decimal | None = None
    due_day: int | None = None
    notes: str | None = None


class LiabilityUpdate(BaseModel):
    type: str | None = None
    name: str | None = None
    institution: str | None = None
    original_amount: Decimal | None = None
    current_balance: Decimal | None = None
    interest_rate: Decimal | None = None
    due_day: int | None = None
    notes: str | None = None
    is_active: bool | None = None


class LiabilityTransactionResponse(ORMModel):
    id: int
    liability_id: int
    transaction_date: date
    type: str
    amount: Decimal
    memo: str | None


class LiabilityTransactionCreate(BaseModel):
    transaction_date: date
    type: str
    amount: Decimal
    memo: str | None = None

File: src/lib/validations/liability.ts
import Decimal from "decimal.js";
import { z } from "zod";

import { dateSchema, decimalSchema } from "@/lib/validations/account";

export const liabilityResponseSchema = z.object({
  id: z.number().int(),
  type: z.string(),
  name: z.string(),
  institution: z.string().nullable(),
  original_amount: decimalSchema,
  current_balance: decimalSchema,
  interest_rate: decimalSchema.nullable(),
  due_day: z.number().int().nullable(),
  notes: z.string().nullable(),
  is_active: z.boolean(),
});

export const liabilityCreateSchema = z.object({
  type: z.string(),
  name: z.string(),
  institution: z.string().nullable().optional(),
  original_amount: decimalSchema.default(new Decimal(0)),
  current_balance: decimalSchema,
  interest_rate: decimalSchema.nullable().optional(),
  due_day: z.number().int().nullable().optional(),
  notes: z.string().nullable().optional(),
});

export const liabilityUpdateSchema = z.object({
  type: z.string().optional(),
  name: z.string().optional(),
  institution: z.string().nullable().optional(),
  original_amount: decimalSchema.optional(),
  current_balance: decimalSchema.optional(),
  interest_rate: decimalSchema.nullable().optional(),
  due_day: z.number().int().nullable().optional(),
  notes: z.string().nullable().optional(),
  is_active: z.boolean().optional(),
});

export const liabilityTransactionResponseSchema = z.object({
  id: z.number().int(),
  liability_id: z.number().int(),
  transaction_date: dateSchema,
  type: z.string(),
  amount: decimalSchema,
  memo: z.string().nullable(),
});

export const liabilityTransactionCreateSchema = z.object({
  transaction_date: dateSchema,
  type: z.string(),
  amount: decimalSchema,
  memo: z.string().nullable().optional(),
});

export type LiabilityResponse = z.infer<typeof liabilityResponseSchema>;
export type LiabilityTransactionResponse = z.infer<typeof liabilityTransactionResponseSchema>;
